File: clojure/kindergarten-garden/src/kindergarten_garden.clj
(ns kindergarten-garden)

(def students [:alice :bob :charlie :david :eve :fred
               :ginny :harriet :ileana :joseph :kincaid :larry])

(def plants {\G :grass \C :clover \R :radishes \V :violets})

(defn garden [s]
  (->> s
       clojure.string/split-lines
       (map #(partition 2 (map plants %)))
       (apply interleave)
       (apply concat)
       (partition 4)
       (zipmap students)))

File: python/kindergarten-garden/kindergarten_garden.py
"""Determine which plants a student is in charge of within a garden"""

# pylint: disable=missing-function-docstring

DEFAULT_STUDENTS = (
    'Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Fred',
    'Ginny', 'Harriet', 'Ileana', 'Joseph', 'Kincaid', 'Larry'
)

PLANTS = {'G': 'Grass', 'C': 'Clover', 'R': 'Radishes', 'V': 'Violets'}


class Garden:  # pylint: disable=missing-class-docstring,too-few-public-methods
    def __init__(self, diagram, students=DEFAULT_STUDENTS):
        self._garden = diagram.split('\n')
        self._students = sorted(students)

    def plants(self, student):
        student_index = self._students.index(student)
        return [
            PLANTS[plant]
            for row in self._garden 
            for plant in row[student_index * 2:student_index * 2 + 2]
        ]
